enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,267 @@
+---
+title: Safe OpenShift Upgrades with Operators 
+authors:
+  - "@awgreene"
+reviewers:
+  - TBD
+  - "@kevinrizza"
+approvers:
+  - "@ecordell"
+  - "@spadgett"
+creation-date: 2021-01-11
+last-updated: 2021-01-19
+status: provisional
+see-also:
+  -  N/A
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# Safe OpenShift Upgrades with Operators
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Definitions
+
+**OLM Managed Operators**: Operators that are installed, managed, and upgraded by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) project.
+
+**Minor OpenShift Upgrade**: When the OpenShift Cluster is upgraded to the next minor version, for example the upgrade from version 4.6 to 4.7. See [Semantic Versioning](https://semver.org/) for more details.
```

Comment:
is it always sequential or is it whatever upgrades OCP supports?

Reply:
At this time OpenShift ALWAYS moves through each Minor OpenShift version when upgrading.

Reply:
there's been some musing about possibly allowing direct 4.6 -> 4.8 hops for folks leaving eus-4.6, but nobody has really dug into that yet.  So while it's generally going to be 4.{y} -> 4.{y+1}, it's good to keep in mind that that may get softened in the future.

Reply:
I don't think OLM can know which next y-stream version the user will pick with the current APIs provided by CVO.

If skipping releases is possible, we likely need to provide some configuration mechanism for OLM (a `targetOCPRelease` squirreled away somewhere). And it will be a docs effort to ensure anyone attempting to skip a release sets that field before update, to give OLM a chance to review the operators it manages.

Has there been any discussion of a two-stage ack for CVO, where the user will set a desired target version but the update won't actually start? That would allow OLM to read the desired target version and adjust accordingly.

Reply:
> Has there been any discussion of a two-stage ack for CVO, where the user will set a desired target version but the update won't actually start? That would allow OLM to read the desired target version and adjust accordingly.

sounds like the preflight checks discussion, but i'm not sure where that stands at this point.

Reply:
@wking the only thing on the table right now is an upgrade that doesn't apply updates to MachineConfigPools and that's dependent on us getting comfortable with kubelet version skew of N-2.

The preflights [enhancement](https://github.com/openshift/enhancements/pull/363) didn't get a warm reception though we do plan to try again in the future. They'll take the form of executing jobs defined in the target release image, so if at some point we were to skip a release it would be a 4.N cluster that fetches the target release image and runs jobs from that image.